In [1]:
#Load the necessary libraries.
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import warnings
import os, sys
import itertools



pd.set_option('Display.max_rows', None)
pd.set_option('Display.max_columns', None)
warnings.filterwarnings('ignore')

Objectve = Sales forecast using time series

In [2]:
df1 = pd.read_csv('../amazon.csv')

df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1465 entries, 0 to 1464
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   product_id           1465 non-null   object
 1   product_name         1465 non-null   object
 2   category             1465 non-null   object
 3   discounted_price     1465 non-null   object
 4   actual_price         1465 non-null   object
 5   discount_percentage  1465 non-null   object
 6   rating               1465 non-null   object
 7   rating_count         1463 non-null   object
 8   about_product        1465 non-null   object
 9   user_id              1465 non-null   object
 10  user_name            1465 non-null   object
 11  review_id            1465 non-null   object
 12  review_title         1465 non-null   object
 13  review_content       1465 non-null   object
 14  img_link             1465 non-null   object
 15  product_link         1465 non-null   object
dtypes: obj

In [3]:
df1.tail()

product_id                                       product_name   
1460  B08L7J3T31  Noir Aqua - 5pcs PP Spun Filter + 1 Spanner | ...  \
1461  B01M6453MB  Prestige Delight PRWO Electric Rice Cooker (1 ...   
1462  B009P2LIL4  Bajaj Majesty RX10 2000 Watts Heat Convector R...   
1463  B00J5DYCCA  Havells Ventil Air DSP 230mm Exhaust Fan (Pist...   
1464  B01486F4G6  Borosil Jumbo 1000-Watt Grill Sandwich Maker (...   

                                               category discounted_price   
1460  Home&Kitchen|Kitchen&HomeAppliances|WaterPurif...             ₹379  \
1461  Home&Kitchen|Kitchen&HomeAppliances|SmallKitch...           ₹2,280   
1462  Home&Kitchen|Heating,Cooling&AirQuality|RoomHe...           ₹2,219   
1463  Home&Kitchen|Heating,Cooling&AirQuality|Fans|E...           ₹1,399   
1464  Home&Kitchen|Kitchen&HomeAppliances|SmallKitch...           ₹2,863   

     actual_price discount_percentage rating rating_count   
1460         ₹919                 59%      4        1,090  \
1461       ₹3,045                 25%    4.1        4,118   
1462       ₹3,080                 28%    3.6          468   
1463       ₹1,890                 26%      4        8,031   
1464       ₹3,690                 22%    4.3        6,987   

                                          about_product   
1460  SUPREME QUALITY 90 GRAM 3 LAYER THIK PP SPUN F...  \
1461                       230 Volts, 400 watts, 1 Year   
1462  International design and styling|Two heat sett...   
1463  Fan sweep area: 230 MM ; Noise level: (40 - 45...   
1464  Brand-Borosil, Specification – 23V ~ 5Hz;1 W...   

                                                user_id   
1460  AHITFY6AHALOFOHOZEOC6XBP4FEA,AFRABBODZJZQB6Z4U...  \
1461  AFG5FM3NEMOL6BNFRV2NK5FNJCHQ,AGEINTRN6Z563RMLH...   
1462  AGVPWCMAHYQWJOQKMUJN4DW3KM5Q,AF4Q3E66MY4SR7YQZ...   
1463  AF2JQCLSCY3QJATWUNNHUSVUPNQQ,AFDMLUXC5LS5RXDJS...   
1464  AFGW5PT3R6ZAVQR4Y5MWVAKBZAYA,AG7QNJ2SCS5VS5VYY...   

                                              user_name   
1460  Prabha ds,Raghuram bk,Real Deal,Amazon Custome...  \
1461  Manu Bhai,Naveenpittu,Evatira Sangma,JAGANNADH...   
1462  Nehal Desai,Danish Parwez,Amazon Customer,Amaz...   
1463  Shubham Dubey,E.GURUBARAN,Mayank S.,eusuf khan...   
1464  Rajib,Ajay B,Vikas Kahol,PARDEEP,Anindya Prama...   

                                              review_id   
1460  R3G3XFHPBFF0E8,R3C0BZCD32EIGW,R2EBVBCN9QPD9R,R...  \
1461  R3DDL2UPKQ2CK9,R2SYYU1OATVIU5,R1VM993161IYRW,R...   
1462  R1TLRJVW4STY5I,R2O455KRN493R1,R3Q5MVGBRIAS2G,R...   
1463  R39Q2Y79MM9SWK,R3079BG1NIH6MB,R29A31ZELTZNJM,R...   
1464  R20RBRZ0WEUJT9,ROKIFK9R2ISSE,R30EEG2FNJSN5I,R2...   

                                           review_title   
1460  Received the product without spanner,Excellent...  \
1461  ok,everything was good couldn't return bcoz I ...   
1462  very good,Work but front melt after 2 month,Go...   
1463  Fan Speed is slow,Good quality,Good product,go...   
1464  Works perfect,Ok good product,Nice Product. Re...   

                                         review_content   
1460  I received product without spanner,Excellent p...  \
1461  ok,got everything as mentioned but the measuri...   
1462  plastic but cool body ,u have to find sturdy s...   
1463  I have installed this in my kitchen working fi...   
1464  It does it job perfectly..only issue is temp c...   

                                               img_link   
1460  https://m.media-amazon.com/images/I/41fDdRtjfx...  \
1461  https://m.media-amazon.com/images/I/41gzDxk4+k...   
1462  https://m.media-amazon.com/images/W/WEBP_40237...   
1463  https://m.media-amazon.com/images/W/WEBP_40237...   
1464  https://m.media-amazon.com/images/W/WEBP_40237...   

                                           product_link  
1460  https://www.amazon.in/Noir-Aqua-Spanner-Purifi...  
1461  https://www.amazon.in/Prestige-Delight-PRWO-1-...  
1462  https://www.amazon.in/Bajaj-RX-10-2000-Watt-Co...  
1463  https://www.amazon.in/Havells-Ve

**Data Wrangling**

In [4]:
df1.columns

Index(['product_id', 'product_name', 'category', 'discounted_price',
       'actual_price', 'discount_percentage', 'rating', 'rating_count',
       'about_product', 'user_id', 'user_name', 'review_id', 'review_title',
       'review_content', 'img_link', 'product_link'],
      dtype='object')

In [5]:
df1[['rating','rating_count']].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1465 entries, 0 to 1464
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   rating        1465 non-null   object
 1   rating_count  1463 non-null   object
dtypes: object(2)
memory usage: 23.0+ KB


In [6]:
df1.isna().sum()

product_id             0
product_name           0
category               0
discounted_price       0
actual_price           0
discount_percentage    0
rating                 0
rating_count           2
about_product          0
user_id                0
user_name              0
review_id              0
review_title           0
review_content         0
img_link               0
product_link           0
dtype: int64

In [7]:
# Apply the Data_cleaning script

sys.path.append('../src')
from Data_cleaning import DataHandling

cleaner = DataHandling(df1)

cleaner.apply_cleaning()

df_cleaned = cleaner.get_cleaned_dataframe()

df_cleaned.head()

Product_id                                       Product_name   
0  B07JW9H4J1  Wayona Nylon Braided USB to Lightning Fast Cha...  \
1  B098NS6PVG  Ambrane Unbreakable 60W / 3A Fast Charging 1.5...   
2  B096MSW6CT  Sounce Fast Phone Charging Cable & Data Sync U...   
3  B08HDJ86NZ  boAt Deuce USB 300 2 in 1 Type-C & Micro USB S...   
4  B08CF3B7N1  Portronics Konnect L 1.2M Fast Charging 3A 8 P...   

                                            Category  Discounted_price   
0  Computers&Accessories|Accessories&Peripherals|...             399.0  \
1  Computers&Accessories|Accessories&Peripherals|...             199.0   
2  Computers&Accessories|Accessories&Peripherals|...             199.0   
3  Computers&Accessories|Accessories&Peripherals|...             329.0   
4  Computers&Accessories|Accessories&Peripherals|...             154.0   

   Actual_price  Discount_percentage  Rating  Rating_count   
0        1099.0                 64.0     4.2       24269.0  \
1         349.0                 43.0     4.0       43994.0   
2        1899.0                 90.0     3.9        7928.0   
3         699.0                 53.0     4.2       94363.0   
4         399.0                 61.0     4.2       16905.0   

                                       About_product   
0  High Compatibility : Compatible With iPhone 12...  \
1  Compatible with all Type C enabled devices, be...   
2  【 Fast Charger& Data Sync】-With built-in safet...   
3  The boAt Deuce USB 300 2 in 1 cable is compati...   
4  [CHARGE & SYNC FUNCTION]- This cable comes wit...   

                                             User_id   
0  AG3D6O4STAQKAY2UVGEUV46KN35Q,AHMY5CWJMMK5BJRBB...  \
1  AECPFYFQVRUWC3KGNLJIOREFP5LQ,AGYYVPDD7YG7FYNBX...   
2  AGU3BBQ2V2DDAMOAKGFAWDDQ6QHA,AESFLDV2PT363T2AQ...   
3  AEWAZDZZJLQUYVOVGBEUKSLXHQ5A,AG5HTSFRRE6NL3M5S...   
4  AE3Q6KSUK5P75D5HFYHCRAOLODSA,AFUGIFH5ZAFXRDSZH...   

                                           User_name   
0  Manav,Adarsh gupta,Sundeep,S.Sayeed Ahmed,jasp...  \
1  ArdKn,Nirbhay kumar,Sagar Viswanathan,Asp,Plac...   
2  Kunal,Himanshu,viswanath,sai niharka,saqib mal...   
3  Omkar dhale,JD,HEMALATHA,Ajwadh a.,amar singh ...   
4  rahuls6099,Swasat Borah,Ajay Wadke,Pranali,RVK...   

                                           Review_id   
0  R3HXWT0LRP0NMF,R2AJM3LFTLZHFO,R6AQJGUP6P86,R1K...  \
1  RGIQEG07R9HS2,R1SMWZQ86XIN8U,R2J3Y1WL29GWDE,RY...   
2  R3J3EQQ9TZI5ZJ,R3E7WBGK7ID0KV,RWU79XKQ6I1QF,R2...   
3  R3EEUZKKK9J36I,R3HJVYCLYOY554,REDECAZ7AMPQC,R1...   
4  R1BP4L2HH9TFUP,R16PVJEXKV6QZS,R2UPDB81N66T4P,R...   

                                        Review_title   
0  Satisfied,Charging is really fast,Value for mo...  \
1  A Good Braided Cable for Your Type C Device,Go...   
2  Good speed for earlier versions,Good Product,W...   
3  Good product,Good one,Nice,Really nice product...   
4  As good as original,Decent,Good one for second...   

                                      Review_content   
0  Looks durable Charging is fine tooNo complains...  \
1  I ordered this cable to connect my phone to An...   
2  Not quite durable and sturdy,https://m.media-a...   
3  Good product,long wire,Charges good,Nice,I bou...   
4  Bought this instead of original apple, does th...   

                                            Img_link   
0  https://m.media-amazon.com/images/W/WEBP_40237...  \
1  https://m.media-amazon.com/images/W/WEBP_40237...   
2  https://m.media-amazon.com/images/W/WEBP_40237...   
3  https://m.media-amazon.com/images/I/41V5FtEWPk...   
4  https://m.media-amazon.com/images/W/WEBP_40237...   

                                        Product_link  
0  https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...  
1  https://www.amazon.in/Ambrane-Unbreakable-Char...  
2  https://www.amazon.in/Sounce-iPhone-Charging-C...  
3  https://www.amazon.in/Deuce-300-Resistant-Tang...  
4  https://www.amazon.in/Portronics-Konnect-POR-1...

In [8]:
#Handle missing values using SimpleImputer
from sklearn.impute import SimpleImputer

Imputer = SimpleImputer(strategy='most_frequent')


df1[['Rating','Rating_count']] = Imputer.fit_transform(df1[['Rating','Rating_count']])

df1[['Rating','Rating_count']].isna().sum()

Rating          0
Rating_count    0
dtype: int64


clean category column, 

In [9]:
# check missing values
df1.isna().sum()


Product_id             0
Product_name           0
Category               0
Discounted_price       0
Actual_price           0
Discount_percentage    0
Rating                 0
Rating_count           0
About_product          0
User_id                0
User_name              0
Review_id              0
Review_title           0
Review_content         0
Img_link               0
Product_link           0
dtype: int64

In [10]:
# rename columns
new_column_name = {
    'Discounted_price':'Discounted_price_$',
    'Actual_price':'Actual_price_$',
    'Discount_percentage':'Discount_percentage_%',
}

df1.rename(columns=new_column_name, inplace=True)

In [11]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1465 entries, 0 to 1464
Data columns (total 16 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Product_id             1465 non-null   object 
 1   Product_name           1465 non-null   object 
 2   Category               1465 non-null   object 
 3   Discounted_price_$     1465 non-null   float64
 4   Actual_price_$         1465 non-null   float64
 5   Discount_percentage_%  1465 non-null   float64
 6   Rating                 1465 non-null   float64
 7   Rating_count           1465 non-null   float64
 8   About_product          1465 non-null   object 
 9   User_id                1465 non-null   object 
 10  User_name              1465 non-null   object 
 11  Review_id              1465 non-null   object 
 12  Review_title           1465 non-null   object 
 13  Review_content         1465 non-null   object 
 14  Img_link               1465 non-null   object 
 15  Prod

In [12]:
#Cleaning categorical columns

df1.select_dtypes(include='object').nunique()

Product_id        1351
Product_name      1337
Category           211
About_product     1293
User_id           1194
User_name         1194
Review_id         1194
Review_title      1194
Review_content    1212
Img_link          1412
Product_link      1465
dtype: int64

**Descriptive Statistics**

In [13]:
num_attributes = df1.select_dtypes(exclude='object')
cat_attributes = df1.select_dtypes(include='object')

**Numerical Attributes**

In [14]:
# Numerical Attributes
describe = num_attributes.describe().T

describe['range'] = (num_attributes.max() - num_attributes.min()).tolist()
describe['variation coefficient'] = (num_attributes.std() / num_attributes.mean()).tolist()
describe['skew'] = num_attributes.skew().tolist()
describe['kurtosis'] = num_attributes.kurtosis().tolist()

describe

count          mean           std   min     25%   
Discounted_price_$     1465.0   3125.310874   6944.304394  39.0   325.0  \
Actual_price_$         1465.0   5444.990635  10874.826864  39.0   800.0   
Discount_percentage_%  1465.0     47.691468     21.635905   0.0    32.0   
Rating                 1465.0      4.096587      0.291574   2.0     4.0   
Rating_count           1465.0  18283.367235  42725.921124   2.0  1191.0   

                          50%      75%       max     range   
Discounted_price_$      799.0   1999.0   77990.0   77951.0  \
Actual_price_$         1650.0   4295.0  139900.0  139861.0   
Discount_percentage_%    50.0     63.0      94.0      94.0   
Rating                    4.1      4.3       5.0       3.0   
Rating_count           5179.0  17325.0  426973.0  426971.0   

                       variation coefficient      skew   kurtosis  
Discounted_price_$                  2.221956  4.452395  25.643590  
Actual_price_$                      1.997217  4.559879  29.723082  
Discount_percentage_%               0.453664 -0.290532  -0.580729  
Rating                              0.071175 -1.243343   4.364576  
Rating_count                        2.336874  5.677990  40.412286

**Categorical Attributes**

In [15]:
# Categorical aatributes

cat_attributes.describe().T


count unique   
Product_id      1465   1351  \
Product_name    1465   1337   
Category        1465    211   
About_product   1465   1293   
User_id         1465   1194   
User_name       1465   1194   
Review_id       1465   1194   
Review_title    1465   1194   
Review_content  1465   1212   
Img_link        1465   1412   
Product_link    1465   1465   

                                                              top freq  
Product_id                                             B07JW9H4J1    3  
Product_name    Fire-Boltt Ninja Call Pro Plus 1.83" Smart Wat...    5  
Category        Computers&Accessories|Accessories&Peripherals|...  233  
About_product   [CHARGE & SYNC FUNCTION]- This cable comes wit...    6  
User_id         AHIKJUDTVJ4T6DV6IUGFYZ5LXMPA,AE55KTFVNXYFD5FPY...   10  
User_name       $@|\|TO$|-|,Sethu madhav,Akash Thakur,Burger P...   10  
Review_id       R3F4T5TRYPTMIG,R3DQIEC603E7AY,R1O4Z15FD40PV5,R...   10  
Review_title    Worked on iPhone 7 and didn’t work on XR,Good ...   10  
Review_content  I am not big on camera usage, personally. I wa...    8  
Img_link        https://m.media-amazon.com/images/I/413sCRKobN...    3  
Product_link    https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...    1

In [16]:
df1.columns

Index(['Product_id', 'Product_name', 'Category', 'Discounted_price_$',
       'Actual_price_$', 'Discount_percentage_%', 'Rating', 'Rating_count',
       'About_product', 'User_id', 'User_name', 'Review_id', 'Review_title',
       'Review_content', 'Img_link', 'Product_link'],
      dtype='object')

In [17]:
cat_attributes.shape

(1465, 11)

In [18]:
#create an only categorical and numerical features dataframe
num_df = df1.select_dtypes(include='float')
cat_df = df1.select_dtypes(include='object')

In [19]:
num_df[:5]

Discounted_price_$  Actual_price_$  Discount_percentage_%  Rating   
0               399.0          1099.0                   64.0     4.2  \
1               199.0           349.0                   43.0     4.0   
2               199.0          1899.0                   90.0     3.9   
3               329.0           699.0                   53.0     4.2   
4               154.0           399.0                   61.0     4.2   

   Rating_count  
0       24269.0  
1       43994.0  
2        7928.0  
3       94363.0  
4       16905.0

**Cleaning text or categorical features**

In [20]:
#categorical dataframe
cat_df[:5]

Product_id                                       Product_name   
0  B07JW9H4J1  Wayona Nylon Braided USB to Lightning Fast Cha...  \
1  B098NS6PVG  Ambrane Unbreakable 60W / 3A Fast Charging 1.5...   
2  B096MSW6CT  Sounce Fast Phone Charging Cable & Data Sync U...   
3  B08HDJ86NZ  boAt Deuce USB 300 2 in 1 Type-C & Micro USB S...   
4  B08CF3B7N1  Portronics Konnect L 1.2M Fast Charging 3A 8 P...   

                                            Category   
0  Computers&Accessories|Accessories&Peripherals|...  \
1  Computers&Accessories|Accessories&Peripherals|...   
2  Computers&Accessories|Accessories&Peripherals|...   
3  Computers&Accessories|Accessories&Peripherals|...   
4  Computers&Accessories|Accessories&Peripherals|...   

                                       About_product   
0  High Compatibility : Compatible With iPhone 12...  \
1  Compatible with all Type C enabled devices, be...   
2  【 Fast Charger& Data Sync】-With built-in safet...   
3  The boAt Deuce USB 300 2 in 1 cable is compati...   
4  [CHARGE & SYNC FUNCTION]- This cable comes wit...   

                                             User_id   
0  AG3D6O4STAQKAY2UVGEUV46KN35Q,AHMY5CWJMMK5BJRBB...  \
1  AECPFYFQVRUWC3KGNLJIOREFP5LQ,AGYYVPDD7YG7FYNBX...   
2  AGU3BBQ2V2DDAMOAKGFAWDDQ6QHA,AESFLDV2PT363T2AQ...   
3  AEWAZDZZJLQUYVOVGBEUKSLXHQ5A,AG5HTSFRRE6NL3M5S...   
4  AE3Q6KSUK5P75D5HFYHCRAOLODSA,AFUGIFH5ZAFXRDSZH...   

                                           User_name   
0  Manav,Adarsh gupta,Sundeep,S.Sayeed Ahmed,jasp...  \
1  ArdKn,Nirbhay kumar,Sagar Viswanathan,Asp,Plac...   
2  Kunal,Himanshu,viswanath,sai niharka,saqib mal...   
3  Omkar dhale,JD,HEMALATHA,Ajwadh a.,amar singh ...   
4  rahuls6099,Swasat Borah,Ajay Wadke,Pranali,RVK...   

                                           Review_id   
0  R3HXWT0LRP0NMF,R2AJM3LFTLZHFO,R6AQJGUP6P86,R1K...  \
1  RGIQEG07R9HS2,R1SMWZQ86XIN8U,R2J3Y1WL29GWDE,RY...   
2  R3J3EQQ9TZI5ZJ,R3E7WBGK7ID0KV,RWU79XKQ6I1QF,R2...   
3  R3EEUZKKK9J36I,R3HJVYCLYOY554,REDECAZ7AMPQC,R1...   
4  R1BP4L2HH9TFUP,R16PVJEXKV6QZS,R2UPDB81N66T4P,R...   

                                        Review_title   
0  Satisfied,Charging is really fast,Value for mo...  \
1  A Good Braided Cable for Your Type C Device,Go...   
2  Good speed for earlier versions,Good Product,W...   
3  Good product,Good one,Nice,Really nice product...   
4  As good as original,Decent,Good one for second...   

                                      Review_content   
0  Looks durable Charging is fine tooNo complains...  \
1  I ordered this cable to connect my phone to An...   
2  Not quite durable and sturdy,https://m.media-a...   
3  Good product,long wire,Charges good,Nice,I bou...   
4  Bought this instead of original apple, does th...   

                                            Img_link   
0  https://m.media-amazon.com/images/W/WEBP_40237...  \
1  https://m.media-amazon.com/images/W/WEBP_40237...   
2  https://m.media-amazon.com/images/W/WEBP_40237...   
3  https://m.media-amazon.com/images/I/41V5FtEWPk...   
4  https://m.media-amazon.com/images/W/WEBP_40237...   

                                        Product_link  
0  https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...  
1  https://www.amazon.in/Ambrane-Unbreakable-Char...  
2  https://www.amazon.in/Sounce-iPhone-Charging-C...  
3  https://www.amazon.in/Deuce-300-Resistant-Tang...  
4  https://www.amazon.in/Portronics-Konnect-POR-1...

Category = '|', user_id to review_title = ',' # clean accordingly

In [21]:
sys.path.append('../src')
from Data_cleaning import CategoricalHandling

cat_cleaner = CategoricalHandling(cat_df)

df2_cleaned = cat_cleaner.apply_cat(lambda x: x.lower() if isinstance(x, str) else x, columns=['User_name'])

df2_cleaned[:5]

Product_id                                       Product_name   
0  B07JW9H4J1  Wayona Nylon Braided USB to Lightning Fast Cha...  \
1  B098NS6PVG  Ambrane Unbreakable 60W / 3A Fast Charging 1.5...   
2  B096MSW6CT  Sounce Fast Phone Charging Cable & Data Sync U...   
3  B08HDJ86NZ  boAt Deuce USB 300 2 in 1 Type-C & Micro USB S...   
4  B08CF3B7N1  Portronics Konnect L 1.2M Fast Charging 3A 8 P...   

                                            Category   
0  Computers&Accessories|Accessories&Peripherals|...  \
1  Computers&Accessories|Accessories&Peripherals|...   
2  Computers&Accessories|Accessories&Peripherals|...   
3  Computers&Accessories|Accessories&Peripherals|...   
4  Computers&Accessories|Accessories&Peripherals|...   

                                       About_product   
0  High Compatibility : Compatible With iPhone 12...  \
1  Compatible with all Type C enabled devices, be...   
2  【 Fast Charger& Data Sync】-With built-in safet...   
3  The boAt Deuce USB 300 2 in 1 cable is compati...   
4  [CHARGE & SYNC FUNCTION]- This cable comes wit...   

                                             User_id   
0  AG3D6O4STAQKAY2UVGEUV46KN35Q,AHMY5CWJMMK5BJRBB...  \
1  AECPFYFQVRUWC3KGNLJIOREFP5LQ,AGYYVPDD7YG7FYNBX...   
2  AGU3BBQ2V2DDAMOAKGFAWDDQ6QHA,AESFLDV2PT363T2AQ...   
3  AEWAZDZZJLQUYVOVGBEUKSLXHQ5A,AG5HTSFRRE6NL3M5S...   
4  AE3Q6KSUK5P75D5HFYHCRAOLODSA,AFUGIFH5ZAFXRDSZH...   

                                           User_name   
0  Manav,Adarsh gupta,Sundeep,S.Sayeed Ahmed,jasp...  \
1  ArdKn,Nirbhay kumar,Sagar Viswanathan,Asp,Plac...   
2  Kunal,Himanshu,viswanath,sai niharka,saqib mal...   
3  Omkar dhale,JD,HEMALATHA,Ajwadh a.,amar singh ...   
4  rahuls6099,Swasat Borah,Ajay Wadke,Pranali,RVK...   

                                           Review_id   
0  R3HXWT0LRP0NMF,R2AJM3LFTLZHFO,R6AQJGUP6P86,R1K...  \
1  RGIQEG07R9HS2,R1SMWZQ86XIN8U,R2J3Y1WL29GWDE,RY...   
2  R3J3EQQ9TZI5ZJ,R3E7WBGK7ID0KV,RWU79XKQ6I1QF,R2...   
3  R3EEUZKKK9J36I,R3HJVYCLYOY554,REDECAZ7AMPQC,R1...   
4  R1BP4L2HH9TFUP,R16PVJEXKV6QZS,R2UPDB81N66T4P,R...   

                                        Review_title   
0  Satisfied,Charging is really fast,Value for mo...  \
1  A Good Braided Cable for Your Type C Device,Go...   
2  Good speed for earlier versions,Good Product,W...   
3  Good product,Good one,Nice,Really nice product...   
4  As good as original,Decent,Good one for second...   

                                      Review_content   
0  Looks durable Charging is fine tooNo complains...  \
1  I ordered this cable to connect my phone to An...   
2  Not quite durable and sturdy,https://m.media-a...   
3  Good product,long wire,Charges good,Nice,I bou...   
4  Bought this instead of original apple, does th...   

                                            Img_link   
0  https://m.media-amazon.com/images/W/WEBP_40237...  \
1  https://m.media-amazon.com/images/W/WEBP_40237...   
2  https://m.media-amazon.com/images/W/WEBP_40237...   
3  https://m.media-amazon.com/images/I/41V5FtEWPk...   
4  https://m.media-amazon.com/images/W/WEBP_40237...   

                                        Product_link  
0  https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...  
1  https://www.amazon.in/Ambrane-Unbreakable-Char...  
2  https://www.amazon.in/Sounce-iPhone-Charging-C...  
3  https://www.amazon.in/Deuce-300-Resistant-Tang...  
4  https://www.amazon.in/Portronics-Konnect-POR-1...

In [22]:
df2_cleaned.shape

(1465, 11)

In [23]:

# splits and flatten a column based on the delimiter , or |
def split_and_flatten(column):
    split_values = column.str.split('[|,]', regex=True).to_list()
    flat_list = list(itertools.chain.from_iterable(split_values))
    return flat_list

#Pad the list to the same size/shape
def pad_list(lists):
    max_len = max(len(lst) for lst in lists)
    return [lst + [None] * (max_len - len(lst)) for lst in lists] 

In [24]:
# global dict to store processed data
flattened_data = {}
#take account of maximum length encountered
max_len = 0

for column in df2_cleaned.columns:
    flattened_column = split_and_flatten(df2_cleaned[column])
    flattened_data[column] = flattened_column
    max_len = max(max_len, len(flattened_column))
    
for key in flattened_data:
    while len(flattened_data[key]) < max_len:
        flattened_data[key].append(None)
    
flattened_data = pd.DataFrame(flattened_data)

for column in flattened_data.columns:
    if flattened_data[column].dtype == object:
        flattened_data[column].fillna('Missing', inplace=True)
    else:
        flattened_data[column].fillna(flattened_data[column].mode()[0], inplace=True)
        
        
flattened_data = flattened_data[~flattened_data.isin(['Missing']).any(axis=1)]



print(flattened_data)

      Product_id                                       Product_name   
0     B07JW9H4J1  Wayona Nylon Braided USB to Lightning Fast Cha...  \
1     B098NS6PVG                                                 12   
2     B096MSW6CT                                                 11   
3     B08HDJ86NZ                                                  X   
4     B08CF3B7N1                                                  8   
5     B08Y1TFSP6                                                  7   
6     B08WRWPM22                                                  6   
7     B08DDRGWTJ                                                  5   
8     B008IFXQFU                                           iPad Air   
9     B082LZGK39                                                Pro   
10    B08CF3D7QR                               Mini (3 FT Pack of 1   
11    B0789LZTCJ                                              Grey)   
12    B07KSMBL2H  Ambrane Unbreakable 60W / 3A Fast Charging 1.5...   
13    

In [25]:
flattened_data.head()

Product_id                                       Product_name   
0  B07JW9H4J1  Wayona Nylon Braided USB to Lightning Fast Cha...  \
1  B098NS6PVG                                                 12   
2  B096MSW6CT                                                 11   
3  B08HDJ86NZ                                                  X   
4  B08CF3B7N1                                                  8   

                  Category                                   About_product   
0    Computers&Accessories  High Compatibility : Compatible With iPhone 12  \
1  Accessories&Peripherals                                              11   
2       Cables&Accessories                                     X/XsMax/Xr    
3                   Cables                                 iPhone 8/8 Plus   
4                USBCables                                 iPhone 7/7 Plus   

                        User_id       User_name       Review_id   
0  AG3D6O4STAQKAY2UVGEUV46KN35Q           Manav  R3HXWT0LRP0NMF  \
1  AHMY5CWJMMK5BJRBBSNLYT3ONILA    Adarsh gupta  R2AJM3LFTLZHFO   
2  AHCTC6ULH4XB6YHDY6PCH2R772LQ         Sundeep    R6AQJGUP6P86   
3  AGYHHIERNXKA6P5T7CZLXKVPT7IQ  S.Sayeed Ahmed  R1KD19VHEDV0OR   
4  AG4OGOFWXJZTQ2HKYIOCOY3KXF2Q  jaspreet singh  R3C02RMYQMK6FC   

              Review_title                                     Review_content   
0                Satisfied     Looks durable Charging is fine tooNo complains  \
1  Charging is really fast                            Charging is really fast   
2          Value for money                                      good product.   
3           Product review               Till now satisfied with the quality.   
4             Good quality  This is a good product . The charging speed is...   

                                            Img_link   
0  https://m.media-amazon.com/images/W/WEBP_40237...  \
1  https://m.media-amazon.com/images/W/WEBP_40237...   
2  https://m.media-amazon.com/images/W/WEBP_40237...   
3  https://m.media-amazon.com/images/I/41V5FtEWPk...   
4  https://m.media-amazon.com/images/W/WEBP_40237...   

                                        Product_link  
0  https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...  
1  https://www.amazon.in/Ambrane-Unbreakable-Char...  
2  https://www.amazon.in/Sounce-iPhone-Charging-C...  
3  https://www.amazon.in/Deuce-300-Resistant-Tang...  
4  https://www.amazon.in/Portronics-Konnect-POR-1...

In [26]:
flattened_data.shape

(1465, 11)

In [27]:
num_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1465 entries, 0 to 1464
Data columns (total 5 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Discounted_price_$     1465 non-null   float64
 1   Actual_price_$         1465 non-null   float64
 2   Discount_percentage_%  1465 non-null   float64
 3   Rating                 1465 non-null   float64
 4   Rating_count           1465 non-null   float64
dtypes: float64(5)
memory usage: 57.4 KB


In [28]:
cat_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1465 entries, 0 to 1464
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Product_id      1465 non-null   object
 1   Product_name    1465 non-null   object
 2   Category        1465 non-null   object
 3   About_product   1465 non-null   object
 4   User_id         1465 non-null   object
 5   User_name       1465 non-null   object
 6   Review_id       1465 non-null   object
 7   Review_title    1465 non-null   object
 8   Review_content  1465 non-null   object
 9   Img_link        1465 non-null   object
 10  Product_link    1465 non-null   object
dtypes: object(11)
memory usage: 126.0+ KB


In [29]:
#merge them together to create one dataframe by creating a new index.
num_df.reset_index(inplace=True)

flattened_data.reset_index(inplace=True)

df2 = pd.merge(num_df,flattened_data, left_on='index', right_on='index', how='outer')

df2.head()

index  Discounted_price_$  Actual_price_$  Discount_percentage_%  Rating   
0      0               399.0          1099.0                   64.0     4.2  \
1      1               199.0           349.0                   43.0     4.0   
2      2               199.0          1899.0                   90.0     3.9   
3      3               329.0           699.0                   53.0     4.2   
4      4               154.0           399.0                   61.0     4.2   

   Rating_count  Product_id   
0       24269.0  B07JW9H4J1  \
1       43994.0  B098NS6PVG   
2        7928.0  B096MSW6CT   
3       94363.0  B08HDJ86NZ   
4       16905.0  B08CF3B7N1   

                                        Product_name                 Category   
0  Wayona Nylon Braided USB to Lightning Fast Cha...    Computers&Accessories  \
1                                                 12  Accessories&Peripherals   
2                                                 11       Cables&Accessories   
3                                                  X                   Cables   
4                                                  8                USBCables   

                                    About_product   
0  High Compatibility : Compatible With iPhone 12  \
1                                              11   
2                                     X/XsMax/Xr    
3                                 iPhone 8/8 Plus   
4                                 iPhone 7/7 Plus   

                        User_id       User_name       Review_id   
0  AG3D6O4STAQKAY2UVGEUV46KN35Q           Manav  R3HXWT0LRP0NMF  \
1  AHMY5CWJMMK5BJRBBSNLYT3ONILA    Adarsh gupta  R2AJM3LFTLZHFO   
2  AHCTC6ULH4XB6YHDY6PCH2R772LQ         Sundeep    R6AQJGUP6P86   
3  AGYHHIERNXKA6P5T7CZLXKVPT7IQ  S.Sayeed Ahmed  R1KD19VHEDV0OR   
4  AG4OGOFWXJZTQ2HKYIOCOY3KXF2Q  jaspreet singh  R3C02RMYQMK6FC   

              Review_title                                     Review_content   
0                Satisfied     Looks durable Charging is fine tooNo complains  \
1  Charging is really fast                            Charging is really fast   
2          Value for money                                      good product.   
3           Product review               Till now satisfied with the quality.   
4             Good quality  This is a good product . The charging speed is...   

                                            Img_link   
0  https://m.media-amazon.com/images/W/WEBP_40237...  \
1  https://m.media-amazon.com/images/W/WEBP_40237...   
2  https://m.media-amazon.com/images/W/WEBP_40237...   
3  https://m.media-amazon.com/images/I/41V5FtEWPk...   
4  https://m.media-amazon.com/images/W/WEBP_40237...   

                                        Product_link  
0  https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...  
1  https://www.amazon.in/Ambrane-Unbreakable-Char...  
2  https://www.amazon.in/Sounce-iPhone-Charging-C...  
3  https://www.amazon.in/Deuce-300-Resistant-Tang...  
4  https://www.amazon.in/Portronics-Konnect-POR-1...

In [30]:

df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1465 entries, 0 to 1464
Data columns (total 17 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   index                  1465 non-null   int64  
 1   Discounted_price_$     1465 non-null   float64
 2   Actual_price_$         1465 non-null   float64
 3   Discount_percentage_%  1465 non-null   float64
 4   Rating                 1465 non-null   float64
 5   Rating_count           1465 non-null   float64
 6   Product_id             1465 non-null   object 
 7   Product_name           1465 non-null   object 
 8   Category               1465 non-null   object 
 9   About_product          1465 non-null   object 
 10  User_id                1465 non-null   object 
 11  User_name              1465 non-null   object 
 12  Review_id              1465 non-null   object 
 13  Review_title           1465 non-null   object 
 14  Review_content         1465 non-null   object 
 15  Img_

In [31]:
#drop the index column and save the dataframe 
if 'index' in df2.columns:
    df2.drop(columns='index', inplace=True)
    
df2.reset_index(drop=True, inplace=True)

df2.to_csv('../data/amazon.csv', index=False)
print('New Dataframe saved as a csv file...')

New Dataframe saved as a csv file...


**Descriptive Statistical analysis on df2**

In [32]:
df2.head()

Discounted_price_$  Actual_price_$  Discount_percentage_%  Rating   
0               399.0          1099.0                   64.0     4.2  \
1               199.0           349.0                   43.0     4.0   
2               199.0          1899.0                   90.0     3.9   
3               329.0           699.0                   53.0     4.2   
4               154.0           399.0                   61.0     4.2   

   Rating_count  Product_id   
0       24269.0  B07JW9H4J1  \
1       43994.0  B098NS6PVG   
2        7928.0  B096MSW6CT   
3       94363.0  B08HDJ86NZ   
4       16905.0  B08CF3B7N1   

                                        Product_name                 Category   
0  Wayona Nylon Braided USB to Lightning Fast Cha...    Computers&Accessories  \
1                                                 12  Accessories&Peripherals   
2                                                 11       Cables&Accessories   
3                                                  X                   Cables   
4                                                  8                USBCables   

                                    About_product   
0  High Compatibility : Compatible With iPhone 12  \
1                                              11   
2                                     X/XsMax/Xr    
3                                 iPhone 8/8 Plus   
4                                 iPhone 7/7 Plus   

                        User_id       User_name       Review_id   
0  AG3D6O4STAQKAY2UVGEUV46KN35Q           Manav  R3HXWT0LRP0NMF  \
1  AHMY5CWJMMK5BJRBBSNLYT3ONILA    Adarsh gupta  R2AJM3LFTLZHFO   
2  AHCTC6ULH4XB6YHDY6PCH2R772LQ         Sundeep    R6AQJGUP6P86   
3  AGYHHIERNXKA6P5T7CZLXKVPT7IQ  S.Sayeed Ahmed  R1KD19VHEDV0OR   
4  AG4OGOFWXJZTQ2HKYIOCOY3KXF2Q  jaspreet singh  R3C02RMYQMK6FC   

              Review_title                                     Review_content   
0                Satisfied     Looks durable Charging is fine tooNo complains  \
1  Charging is really fast                            Charging is really fast   
2          Value for money                                      good product.   
3           Product review               Till now satisfied with the quality.   
4             Good quality  This is a good product . The charging speed is...   

                                            Img_link   
0  https://m.media-amazon.com/images/W/WEBP_40237...  \
1  https://m.media-amazon.com/images/W/WEBP_40237...   
2  https://m.media-amazon.com/images/W/WEBP_40237...   
3  https://m.media-amazon.com/images/I/41V5FtEWPk...   
4  https://m.media-amazon.com/images/W/WEBP_40237...   

                                        Product_link  
0  https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...  
1  https://www.amazon.in/Ambrane-Unbreakable-Char...  
2  https://www.amazon.in/Sounce-iPhone-Charging-C...  
3  https://www.amazon.in/Deuce-300-Resistant-Tang...  
4  https://www.amazon.in/Portronics-Konnect-POR-1...

In [33]:
#sort the dataset by their features

num_attributes = df2.select_dtypes(exclude='object')
cat_attributes = df2.select_dtypes(include='object')


**Numerical Attributes**

In [34]:
describe = num_attributes.describe().T

describe['range'] = (num_attributes.max() - num_attributes.min()).tolist()
describe['variation coefficient'] = (num_attributes.std() / num_attributes.mean()).tolist()
describe['skew'] = num_attributes.skew().tolist()
describe['kurtosis'] = num_attributes.kurtosis().tolist()


describe

count          mean           std   min     25%   
Discounted_price_$     1465.0   3125.310874   6944.304394  39.0   325.0  \
Actual_price_$         1465.0   5444.990635  10874.826864  39.0   800.0   
Discount_percentage_%  1465.0     47.691468     21.635905   0.0    32.0   
Rating                 1465.0      4.096587      0.291574   2.0     4.0   
Rating_count           1465.0  18283.367235  42725.921124   2.0  1191.0   

                          50%      75%       max     range   
Discounted_price_$      799.0   1999.0   77990.0   77951.0  \
Actual_price_$         1650.0   4295.0  139900.0  139861.0   
Discount_percentage_%    50.0     63.0      94.0      94.0   
Rating                    4.1      4.3       5.0       3.0   
Rating_count           5179.0  17325.0  426973.0  426971.0   

                       variation coefficient      skew   kurtosis  
Discounted_price_$                  2.221956  4.452395  25.643590  
Actual_price_$                      1.997217  4.559879  29.723082  
Discount_percentage_%               0.453664 -0.290532  -0.580729  
Rating                              0.071175 -1.243343   4.364576  
Rating_count                        2.336874  5.677990  40.412286


Descriptive Statistics Report: Numerical Data\
This report presents an analysis of key descriptive statistics for a dataset containing information on prices, discounts, ratings, and rating counts for 1,465 products. The variables analyzed are Discounted Price ($), Actual Price ($), Discount Percentage (%), Rating, and Rating Count. Each variable's distribution is summarized through statistics such as the mean, standard deviation, skewness, kurtosis, and more.

1. Discounted Price ($)
Mean (Average): $3,125.31
Standard Deviation: $6,944.30
Minimum Value: $39.00
Maximum Value: $77,990.00
Range: $77,951.00
25th Percentile (Q1): $325.00
50th Percentile (Median): $799.00
75th Percentile (Q3): $1,999.00
Coefficient of Variation: 2.22 (indicating high relative variability)
Skewness: 4.45 (indicating a strong positive skew, with a long right tail)
Kurtosis: 25.64 (indicating heavy tails and a peaked distribution)
Interpretation: The discounted prices are highly variable, with most prices being lower but a few extremely high prices driving up the mean. The positive skew suggests that most products are priced relatively low, but there are some high-priced outliers. The high kurtosis further confirms the presence of these outliers.

2. Actual Price ($)
Mean (Average): $5,444.99
Standard Deviation: $10,874.83
Minimum Value: $39.00
Maximum Value: $139,900.00
Range: $139,861.00
25th Percentile (Q1): $800.00
50th Percentile (Median): $1,650.00
75th Percentile (Q3): $4,295.00
Coefficient of Variation: 2.00
Skewness: 4.56 (indicating a strong positive skew)
Kurtosis: 29.72 (indicating very heavy tails)
Interpretation: Actual prices are even more dispersed than discounted prices, with a mean significantly higher than the median. The data has a positive skew, with a majority of lower-priced items and some very high-priced ones pulling the average up. The extreme kurtosis indicates an even greater presence of outliers than in the discounted prices.

3. Discount Percentage (%)
Mean (Average): 47.69%
Standard Deviation: 21.64%
Minimum Value: 0%
Maximum Value: 94%
Range: 94%
25th Percentile (Q1): 32%
50th Percentile (Median): 50%
75th Percentile (Q3): 63%
Coefficient of Variation: 0.45
Skewness: -0.29 (indicating slight negative skew)
Kurtosis: -0.58 (indicating slightly lighter tails)
Interpretation: Discount percentages are fairly normally distributed around a mean of roughly 48%, with most discounts lying between 32% and 63%. The negative skew suggests a slight tendency towards higher discounts, while the negative kurtosis indicates a more uniform distribution without extreme outliers.

4. Rating
Mean (Average): 4.10
Standard Deviation: 0.29
Minimum Value: 2.0
Maximum Value: 5.0
Range: 3.0
25th Percentile (Q1): 4.0
50th Percentile (Median): 4.1
75th Percentile (Q3): 4.3
Coefficient of Variation: 0.07
Skewness: -1.24 (indicating a negative skew)
Kurtosis: 4.36 (indicating a peaked distribution)
Interpretation: Product ratings are tightly clustered around the mean of 4.10, with most products having a rating between 4.0 and 4.3. The slight negative skew suggests that ratings are slightly skewed towards the higher end, with relatively few low-rated products. The positive kurtosis implies a concentration of ratings around the mean, with fewer deviations.

5. Rating Count
Mean (Average): 18,283.37
Standard Deviation: 42,725.92
Minimum Value: 2
Maximum Value: 426,973
Range: 426,971
25th Percentile (Q1): 1,191
50th Percentile (Median): 5,179
75th Percentile (Q3): 17,325
Coefficient of Variation: 2.34
Skewness: 5.68 (indicating strong positive skew)
Kurtosis: 40.41 (indicating extremely heavy tails)
Interpretation: Rating counts vary widely, with the median count significantly lower than the mean, suggesting that while most products have a relatively small number of ratings, a few have an exceptionally high number. This is confirmed by the high positive skew and extreme kurtosis, indicating the presence of very large outliers.

Summary
The dataset exhibits a considerable amount of variability, particularly in prices and rating counts, with a strong tendency for positive skewness and high kurtosis in these variables, indicating the presence of outliers and extreme values. On the other hand, rating scores and discount percentages are more stable and normally distributed. These characteristics suggest that while most products are priced and rated within a certain range, there are a few exceptional products that significantly affect the overall distribution.

**Categorical Attributes**

In [35]:
cat_attributes.describe().T

count unique   
Product_id      1465   1351  \
Product_name    1465   1049   
Category        1465     36   
About_product   1465   1045   
User_id         1465   1102   
User_name       1465    983   
Review_id       1465   1105   
Review_title    1465    883   
Review_content  1465   1020   
Img_link        1465   1412   
Product_link    1465   1465   

                                                              top freq  
Product_id                                             B07JW9H4J1    3  
Product_name                                               Black)   29  
Category                                                   Cables  168  
About_product                                     gaming console    11  
User_id                              AG3D6O4STAQKAY2UVGEUV46KN35Q    5  
User_name                                         Amazon Customer   83  
Review_id                                          R3HXWT0LRP0NMF    5  
Review_title                                                 Good   85  
Review_content                                               Good   31  
Img_link        https://m.media-amazon.com/images/I/413sCRKobN...    3  
Product_link    https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...    1

Descriptive Statistics Report: Categorical Data\
This report analyzes the categorical variables within a dataset of 1,465 product reviews. The variables include Product ID, Product Name, Category, About Product, User ID, User Name, Review ID, Review Title, Review Content, Image Link, and Product Link. Each variable is summarized by the number of unique values, the most frequent value (mode), and the frequency of that mode.

1. Product ID
Count: 1,465
Unique Values: 1,351
Most Frequent Value (Top): B07JW9H4J1
Frequency of Most Frequent Value (Freq): 3
Interpretation: There are 1,351 unique product IDs in the dataset, with one particular product (B07JW9H4J1) appearing three times. This suggests that most products are reviewed only once or twice, with a few exceptions.

2. Product Name
Count: 1,465
Unique Values: 1,049
Most Frequent Value (Top): Black)
Frequency of Most Frequent Value (Freq): 29
Interpretation: There are 1,049 unique product names. The name "Black)" appears most frequently, 29 times. This could indicate a common naming pattern or a generic description often repeated across similar products.

3. Category
Count: 1,465
Unique Values: 36
Most Frequent Value (Top): Cables
Frequency of Most Frequent Value (Freq): 168
Interpretation: The dataset contains 36 different categories, with "Cables" being the most common, appearing in 168 records. This suggests that a significant portion of the products reviewed belong to this category.

4. About Product
Count: 1,465
Unique Values: 1,045
Most Frequent Value (Top): gaming console
Frequency of Most Frequent Value (Freq): 11
Interpretation: There are 1,045 unique descriptions in the "About Product" field, with "gaming console" being mentioned 11 times. This indicates a wide variety of product descriptions, though some terms like "gaming console" are repeated across several products.

5. User ID
Count: 1,465
Unique Values: 1,102
Most Frequent Value (Top): AG3D6O4STAQKAY2UVGEUV46KN35Q
Frequency of Most Frequent Value (Freq): 5
Interpretation: The dataset has 1,102 unique user IDs, with the most frequent user ID appearing 5 times. This suggests that while most reviews are from different users, a few users have left multiple reviews.

6. User Name
Count: 1,465
Unique Values: 983
Most Frequent Value (Top): Amazon Customer
Frequency of Most Frequent Value (Freq): 83
Interpretation: The "User Name" field has 983 unique entries. "Amazon Customer" is the most common user name, appearing 83 times, likely because it's a default or anonymous identifier used by many users.

7. Review ID
Count: 1,465
Unique Values: 1,105
Most Frequent Value (Top): R3HXWT0LRP0NMF
Frequency of Most Frequent Value (Freq): 5
Interpretation: There are 1,105 unique review IDs. The most frequent review ID appears 5 times, indicating that while most reviews have unique identifiers, a few are duplicated.

8. Review Title
Count: 1,465
Unique Values: 883
Most Frequent Value (Top): Good
Frequency of Most Frequent Value (Freq): 85
Interpretation: The dataset contains 883 unique review titles, with "Good" being the most common, appearing 85 times. This suggests that many users opt for brief, positive titles in their reviews.

9. Review Content
Count: 1,465
Unique Values: 1,020
Most Frequent Value (Top): Good
Frequency of Most Frequent Value (Freq): 31
Interpretation: There are 1,020 unique pieces of review content. Similar to the review titles, the word "Good" is the most frequent review content, appearing 31 times, indicating a tendency towards brief and positive feedback.

10. Image Link
Count: 1,465
Unique Values: 1,412
Most Frequent Value (Top): https://m.media-amazon.com/images/I/413sCRKobN...
Frequency of Most Frequent Value (Freq): 3
Interpretation: The dataset has 1,412 unique image links, with one specific image link appearing 3 times. This suggests that most products have unique images, though a few may share the same image link.

11. Product Link
Count: 1,465
Unique Values: 1,465
Most Frequent Value (Top): N/A
Frequency of Most Frequent Value (Freq): 1
Interpretation: Every product has a unique link, with no repetition. This implies that each product is distinctly referenced by its own URL.

Summary
The dataset is characterized by a high level of uniqueness across most variables, with only a few values recurring multiple times. This suggests that while the products and reviews cover a broad range, there are some common elements such as the frequent use of the review title and content "Good", and the category "Cables". Additionally, "Amazon Customer" is a prevalent user identifier, reflecting anonymous or default user entries.

## Business Analytics

In [36]:
df2.columns

Index(['Discounted_price_$', 'Actual_price_$', 'Discount_percentage_%',
       'Rating', 'Rating_count', 'Product_id', 'Product_name', 'Category',
       'About_product', 'User_id', 'User_name', 'Review_id', 'Review_title',
       'Review_content', 'Img_link', 'Product_link'],
      dtype='object')

In [37]:
df2.head()

Discounted_price_$  Actual_price_$  Discount_percentage_%  Rating   
0               399.0          1099.0                   64.0     4.2  \
1               199.0           349.0                   43.0     4.0   
2               199.0          1899.0                   90.0     3.9   
3               329.0           699.0                   53.0     4.2   
4               154.0           399.0                   61.0     4.2   

   Rating_count  Product_id   
0       24269.0  B07JW9H4J1  \
1       43994.0  B098NS6PVG   
2        7928.0  B096MSW6CT   
3       94363.0  B08HDJ86NZ   
4       16905.0  B08CF3B7N1   

                                        Product_name                 Category   
0  Wayona Nylon Braided USB to Lightning Fast Cha...    Computers&Accessories  \
1                                                 12  Accessories&Peripherals   
2                                                 11       Cables&Accessories   
3                                                  X                   Cables   
4                                                  8                USBCables   

                                    About_product   
0  High Compatibility : Compatible With iPhone 12  \
1                                              11   
2                                     X/XsMax/Xr    
3                                 iPhone 8/8 Plus   
4                                 iPhone 7/7 Plus   

                        User_id       User_name       Review_id   
0  AG3D6O4STAQKAY2UVGEUV46KN35Q           Manav  R3HXWT0LRP0NMF  \
1  AHMY5CWJMMK5BJRBBSNLYT3ONILA    Adarsh gupta  R2AJM3LFTLZHFO   
2  AHCTC6ULH4XB6YHDY6PCH2R772LQ         Sundeep    R6AQJGUP6P86   
3  AGYHHIERNXKA6P5T7CZLXKVPT7IQ  S.Sayeed Ahmed  R1KD19VHEDV0OR   
4  AG4OGOFWXJZTQ2HKYIOCOY3KXF2Q  jaspreet singh  R3C02RMYQMK6FC   

              Review_title                                     Review_content   
0                Satisfied     Looks durable Charging is fine tooNo complains  \
1  Charging is really fast                            Charging is really fast   
2          Value for money                                      good product.   
3           Product review               Till now satisfied with the quality.   
4             Good quality  This is a good product . The charging speed is...   

                                            Img_link   
0  https://m.media-amazon.com/images/W/WEBP_40237...  \
1  https://m.media-amazon.com/images/W/WEBP_40237...   
2  https://m.media-amazon.com/images/W/WEBP_40237...   
3  https://m.media-amazon.com/images/I/41V5FtEWPk...   
4  https://m.media-amazon.com/images/W/WEBP_40237...   

                                        Product_link  
0  https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...  
1  https://www.amazon.in/Ambrane-Unbreakable-Char...  
2  https://www.amazon.in/Sounce-iPhone-Charging-C...  
3  https://www.amazon.in/Deuce-300-Resistant-Tang...  
4  https://www.amazon.in/Portronics-Konnect-POR-1...

**Sale analysis**

In [38]:
# Revenue calculation

#calculate the revenue for each product
df2["Revenue_$"] = df2['Discount_percentage_%'] * df2['Rating_count']

#calculate the revenue for all products
total_revenue = df2['Revenue_$'].sum()

print(f'Total Revenue: ${total_revenue}')

df2.head()

Total Revenue: $1292691382.0


Discounted_price_$  Actual_price_$  Discount_percentage_%  Rating   
0               399.0          1099.0                   64.0     4.2  \
1               199.0           349.0                   43.0     4.0   
2               199.0          1899.0                   90.0     3.9   
3               329.0           699.0                   53.0     4.2   
4               154.0           399.0                   61.0     4.2   

   Rating_count  Product_id   
0       24269.0  B07JW9H4J1  \
1       43994.0  B098NS6PVG   
2        7928.0  B096MSW6CT   
3       94363.0  B08HDJ86NZ   
4       16905.0  B08CF3B7N1   

                                        Product_name                 Category   
0  Wayona Nylon Braided USB to Lightning Fast Cha...    Computers&Accessories  \
1                                                 12  Accessories&Peripherals   
2                                                 11       Cables&Accessories   
3                                                  X                   Cables   
4                                                  8                USBCables   

                                    About_product   
0  High Compatibility : Compatible With iPhone 12  \
1                                              11   
2                                     X/XsMax/Xr    
3                                 iPhone 8/8 Plus   
4                                 iPhone 7/7 Plus   

                        User_id       User_name       Review_id   
0  AG3D6O4STAQKAY2UVGEUV46KN35Q           Manav  R3HXWT0LRP0NMF  \
1  AHMY5CWJMMK5BJRBBSNLYT3ONILA    Adarsh gupta  R2AJM3LFTLZHFO   
2  AHCTC6ULH4XB6YHDY6PCH2R772LQ         Sundeep    R6AQJGUP6P86   
3  AGYHHIERNXKA6P5T7CZLXKVPT7IQ  S.Sayeed Ahmed  R1KD19VHEDV0OR   
4  AG4OGOFWXJZTQ2HKYIOCOY3KXF2Q  jaspreet singh  R3C02RMYQMK6FC   

              Review_title                                     Review_content   
0                Satisfied     Looks durable Charging is fine tooNo complains  \
1  Charging is really fast                            Charging is really fast   
2          Value for money                                      good product.   
3           Product review               Till now satisfied with the quality.   
4             Good quality  This is a good product . The charging speed is...   

                                            Img_link   
0  https://m.media-amazon.com/images/W/WEBP_40237...  \
1  https://m.media-amazon.com/images/W/WEBP_40237...   
2  https://m.media-amazon.com/images/W/WEBP_40237...   
3  https://m.media-amazon.com/images/I/41V5FtEWPk...   
4  https://m.media-amazon.com/images/W/WEBP_40237...   

                                        Product_link  Revenue_$  
0  https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...  1553216.0  
1  https://www.amazon.in/Ambrane-Unbreakable-Char...  1891742.0  
2  https://www.amazon.in/Sounce-iPhone-Charging-C...   713520.0  
3  https://www.amazon.in/Deuce-300-Resistant-Tang...  5001239.0  
4  https://www.amazon.in/Portronics-Konnect-POR-1...  1031205.0

In [39]:
# Top Product: Best selling product by revenue
top_prod_sorted = df2.sort_values(by='Revenue_$', ascending=False)

Top_product = top_prod_sorted

print(f"Top Best-Selling Product by Revenue:\n {Top_product[['Product_id', 'Product_name', 'Revenue_$']][:6]}") 

Top Best-Selling Product by Revenue:
      Product_id                                       Product_name   Revenue_$
65   B014I8SX4Y                     Laptops & Other Type C Devices  33303894.0
12   B07KSMBL2H  Ambrane Unbreakable 60W / 3A Fast Charging 1.5...  29461137.0
684  B07KSMBL2H  7SEVEN® Compatible Vu Smart Tv Remote Control ...  29461068.0
352  B07GPXXNNG   Belkin Apple Certified Lightning To Type C Cable  23641345.0
584  B071Z8M4KX                           Compatible with TV Stick  22913793.0
400  B07GQD4K6L                                              Linux  22550206.0


In [40]:
# Top Category: Which Product categories generate the most revenue

#group the category by the sum of revenue
category_revenue = df2.groupby('Category')[['Product_name','Revenue_$']].sum().reset_index()
#sort the category 
category_revenue_sorted = category_revenue.sort_values(by='Revenue_$', ascending=False)
#identify the top categories
top_categories = category_revenue_sorted

print(f"Top Product categories by Revenue: \n {top_categories.head(5)}")


Top Product categories by Revenue: 
                    Category   
6     Computers&Accessories  \
5        Cables&Accessories   
4                    Cables   
2   Accessories&Peripherals   
34                USBCables   

                                         Product_name    Revenue_$  
6   Wayona Nylon Braided USB to Lightning Fast Cha...  208536408.0  
5   11 5Ambrane Unbreakable 60W / 3A Fast Charging...  175634812.0  
4    X iPad Air Tablets Black) 8 Pro 10000+ Bends ...  170042189.0  
2    12 6 Grey) 480Mbps Data Sync11 5 Tangle-Freep...  121565309.0  
34   8 Pro Laptops & other Type C devicesSounce Fa...  115753951.0  


**Customer Analysis**

In [41]:
from datetime import datetime, timedelta

# create a possible date for purchase date

start_date = datetime(2021,1,1)
end_date = datetime(2023, 12, 31)

def random_date(start, end):
    """Generate random datetime between start and end."""
    return start + timedelta(days=np.random.randint(0, (end - start).days))

df2['Purchase_date'] = df2.apply(lambda x: random_date(start_date, end_date), axis=1)




In [42]:
# Customer segmentation based on their purchase behaviour

#confirm the datetime preparation
today = datetime.now()
df2['Purchase_date'] = pd.to_datetime(df2['Purchase_date'])

#Calculate Recency, Frequency, and Monetary values for each customer
rfm = df2.groupby('User_id').agg({
    'Purchase_date': lambda x: (today - x.max()).days,
    'User_id': 'count',
    'Discounted_price_$': 'sum'
})

#rename columns
rfm.columns = ['Recency', 'Frequency', 'Monetary']

rfm = rfm.reset_index()

#rank customers based on RFM values
rfm['R_rank'] = rfm['Recency'].rank(ascending=True)
rfm['F_rank'] = rfm['Frequency'].rank(ascending=False)
rfm['M_rank'] = rfm['Monetary'].rank(ascending=False)

#create a combined RFM score by summing the ranks
rfm['RFM_Score'] = rfm['R_rank'] + rfm['F_rank'] + rfm['M_rank']

#Segment customers based on the RFM score
rfm['Customer_Segment'] = pd.qcut(rfm['RFM_Score'].astype(int), q=4, labels=['Low', 'Medium','High', 'VIP'])

rfm.head(10)

User_id  Recency  Frequency  Monetary  R_rank  F_rank   
0  AE242TR3GQ6TYC6W4SJ5UYYKBTYQ     1277          1     999.0  1046.5   668.5  \
1  AE2E632GMYL5U2ESNXOX5UT5D34A     1305          1     299.0  1064.5   668.5   
2  AE2JTMRKTUOIVIZWS2WDGTMNTU4Q     1020          2    4498.0   824.0   162.0   
3  AE2ODWBBOBD2SITDDIEJ644OSRFQ      514          2     358.0   301.0   162.0   
4  AE2PQX6JEE6UW7QB6SNEYP3TAXLQ     1103          1    1519.0   919.5   668.5   
5  AE2PYPC3OF2HEF4NCE63FBFFFOMQ     1101          1   19999.0   916.0   668.5   
6  AE2SWHFUXPXFJLITBNYV5YGTI5LQ      471          1     899.0   262.0   668.5   
7  AE2THTCCQLBIUSWPF4CPXC6GGP7Q      569          2   24598.0   373.5   162.0   
8  AE2VXY4CFO36MDSIMPG43XHNF4GA      314          3   14699.0    60.0    57.0   
9  AE2YKXGI2XFOVDHNL6FF2RQAZ55A     1326          1    2249.0  1080.5   668.5   

   M_rank  RFM_Score Customer_Segment  
0   602.0     2317.0              VIP  
1   913.5     2646.5              VIP  
2   221.0     1207.0              Low  
3   864.0     1327.0           Medium  
4   467.5     2055.5             High  
5    57.0     1641.5           Medium  
6   634.0     1564.5           Medium  
7    42.0      577.5              Low  
8    87.0      204.0              Low  
9   362.5     2111.5             High

In [43]:
rfm.to_csv('../data/cleaned/rfm.csv', index=False)

print(rfm.shape)

(1102, 9)


**Price Analysis**

In [44]:
import scipy
from scipy.stats import ttest_ind, f_oneway

#calculating additional metric

df2['Discount_amount'] = df2['Actual_price_$'] - df2['Discounted_price_$']
df2['Sales_volume'] = df2['Rating_count'] #using this as proxy for sales volumne

#categorize discount into bins
discount_bins = [0, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100]
discount_labels = ['0-10%', '10-20%', '20-30%', '30-40%', '40-50%', '50-60%', '60-70%', '70-80%', '80-90%', '90-100%']
df2['Discount_bin'] = pd.cut(df2['Discount_percentage_%'], bins=discount_bins, labels=discount_labels, include_lowest=True)

#Group the discount bin and calculate average sales volumne
discount_effectiveness = df2.groupby('Discount_bin').agg({
    'Sales_volume':['mean','sum'],
    'Discount_percentage_%':'count'
}).reset_index()

#Rename the columns
discount_effectiveness.columns = ['Discount_bin', 'Avg_Sales_Volume', 'Total_Sales_Volume', 'Count']

# Statistical test to compare sales volumes across discount levels
no_discount_sales = df2[df2['Discount_percentage_%'] == 0]['Sales_volume']
discounted_sales = df2[df2['Discount_percentage_%'] > 0]['Sales_volume']

# T-test for two independent samples
t_stat, p_val = ttest_ind(no_discount_sales, discounted_sales, equal_var=False)
print(f"T-test results: t-statistic = {t_stat}, p-value = {p_val}")

# comparing more than two groups (ANOVA)
anova_results = f_oneway(*[df2[df2['Discount_bin'] == bin]['Sales_volume'] for bin in discount_labels if bin in df2['Discount_bin'].values])
print(f"ANOVA results: F-statistic = {anova_results.statistic}, p-value = {anova_results.pvalue}")

discount_effectiveness.head()

T-test results: t-statistic = -5.431926319808806, p-value = 2.2009550084717845e-07
ANOVA results: F-statistic = 2.732597703877533, p-value = 0.0036486604214584824


Discount_bin  Avg_Sales_Volume  Total_Sales_Volume  Count
0        0-10%      12086.261905           1015246.0     84
1       10-20%      16486.489583           1582703.0     96
2       20-30%      24242.048780           3975696.0    164
3       30-40%      16517.655172           2874072.0    174
4       40-50%      16739.071429           4218246.0    252

**Price Elasticity**

In [45]:
#Calculate percentage change in price and quantity demanded
df2['Price_change_%'] = df2['Actual_price_$'].pct_change() * 100
df2['Quantity_change_%'] = df2['Sales_volume'].pct_change() * 100

#Handle NaN values 
df2[['Price_change_%','Quantity_change_%']] = df2[['Price_change_%','Quantity_change_%']].fillna(df2[['Price_change_%','Quantity_change_%']].mean())

#Calculate Price Elasticity of Demand (PED)
df2['PED'] = df2['Quantity_change_%'] / df2['Price_change_%']

print(df2[['Product_id', 'Actual_price_$', 'Sales_volume', 'Price_change_%', 'Quantity_change_%', 'PED']][:5])

   Product_id  Actual_price_$  Sales_volume  Price_change_%   
0  B07JW9H4J1          1099.0       24269.0      462.267553  \
1  B098NS6PVG           349.0       43994.0      -68.243858   
2  B096MSW6CT          1899.0        7928.0      444.126074   
3  B08HDJ86NZ           699.0       94363.0      -63.191153   
4  B08CF3B7N1           399.0       16905.0      -42.918455   

   Quantity_change_%        PED  
0        2103.264801   4.549886  
1          81.276526  -1.190972  
2         -81.979361  -0.184586  
3        1090.249748 -17.253202  
4         -82.085139   1.912584  


In [46]:
df2.head()

Discounted_price_$  Actual_price_$  Discount_percentage_%  Rating   
0               399.0          1099.0                   64.0     4.2  \
1               199.0           349.0                   43.0     4.0   
2               199.0          1899.0                   90.0     3.9   
3               329.0           699.0                   53.0     4.2   
4               154.0           399.0                   61.0     4.2   

   Rating_count  Product_id   
0       24269.0  B07JW9H4J1  \
1       43994.0  B098NS6PVG   
2        7928.0  B096MSW6CT   
3       94363.0  B08HDJ86NZ   
4       16905.0  B08CF3B7N1   

                                        Product_name                 Category   
0  Wayona Nylon Braided USB to Lightning Fast Cha...    Computers&Accessories  \
1                                                 12  Accessories&Peripherals   
2                                                 11       Cables&Accessories   
3                                                  X                   Cables   
4                                                  8                USBCables   

                                    About_product   
0  High Compatibility : Compatible With iPhone 12  \
1                                              11   
2                                     X/XsMax/Xr    
3                                 iPhone 8/8 Plus   
4                                 iPhone 7/7 Plus   

                        User_id       User_name       Review_id   
0  AG3D6O4STAQKAY2UVGEUV46KN35Q           Manav  R3HXWT0LRP0NMF  \
1  AHMY5CWJMMK5BJRBBSNLYT3ONILA    Adarsh gupta  R2AJM3LFTLZHFO   
2  AHCTC6ULH4XB6YHDY6PCH2R772LQ         Sundeep    R6AQJGUP6P86   
3  AGYHHIERNXKA6P5T7CZLXKVPT7IQ  S.Sayeed Ahmed  R1KD19VHEDV0OR   
4  AG4OGOFWXJZTQ2HKYIOCOY3KXF2Q  jaspreet singh  R3C02RMYQMK6FC   

              Review_title                                     Review_content   
0                Satisfied     Looks durable Charging is fine tooNo complains  \
1  Charging is really fast                            Charging is really fast   
2          Value for money                                      good product.   
3           Product review               Till now satisfied with the quality.   
4             Good quality  This is a good product . The charging speed is...   

                                            Img_link   
0  https://m.media-amazon.com/images/W/WEBP_40237...  \
1  https://m.media-amazon.com/images/W/WEBP_40237...   
2  https://m.media-amazon.com/images/W/WEBP_40237...   
3  https://m.media-amazon.com/images/I/41V5FtEWPk...   
4  https://m.media-amazon.com/images/W/WEBP_40237...   

                                        Product_link  Revenue_$ Purchase_date   
0  https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...  1553216.0    2022-11-10  \
1  https://www.amazon.in/Ambrane-Unbreakable-Char...  1891742.0    2023-07-02   
2  https://www.amazon.in/Sounce-iPhone-Charging-C...   713520.0    2021-10-05   
3  https://www.amazon.in/Deuce-300-Resistant-Tang...  5001239.0    2023-12-22   
4  https://www.amazon.in/Portronics-Konnect-POR-1...  1031205.0    2023-06-18   

   Discount_amount  Sales_volume Discount_bin  Price_change_%   
0            700.0       24269.0       60-70%      462.267553  \
1            150.0       43994.0       40-50%      -68.243858   
2           1700.0        7928.0       80-90%      444.126074   
3            370.0       94363.0       50-60%      -63.191153   
4            245.0       16905.0       60-70%      -42.918455   

   Quantity_change_%        PED  
0        2103.264801   4.549886  
1          81.276526  -1.190972  
2         -81.979361  -0.184586  
3        1090.249748 -17.253202  
4         -82.085139   1.912584

In [47]:
df2.isna().sum()

Discounted_price_$       0
Actual_price_$           0
Discount_percentage_%    0
Rating                   0
Rating_count             0
Product_id               0
Product_name             0
Category                 0
About_product            0
User_id                  0
User_name                0
Review_id                0
Review_title             0
Review_content           0
Img_link                 0
Product_link             0
Revenue_$                0
Purchase_date            0
Discount_amount          0
Sales_volume             0
Discount_bin             0
Price_change_%           0
Quantity_change_%        0
PED                      2
dtype: int64

In [48]:
#fix the NaN in PED with the median
df2['PED'] = df2['PED'].fillna(df2['PED'].median())
#check for NaN values
df2.isna().sum()

Discounted_price_$       0
Actual_price_$           0
Discount_percentage_%    0
Rating                   0
Rating_count             0
Product_id               0
Product_name             0
Category                 0
About_product            0
User_id                  0
User_name                0
Review_id                0
Review_title             0
Review_content           0
Img_link                 0
Product_link             0
Revenue_$                0
Purchase_date            0
Discount_amount          0
Sales_volume             0
Discount_bin             0
Price_change_%           0
Quantity_change_%        0
PED                      0
dtype: int64

In [49]:
# get the selling price

new_column = {
    "Discount_amount":"Selling_price_$"
}

df2.rename(columns=new_column, inplace=True)

In [50]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1465 entries, 0 to 1464
Data columns (total 24 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   Discounted_price_$     1465 non-null   float64       
 1   Actual_price_$         1465 non-null   float64       
 2   Discount_percentage_%  1465 non-null   float64       
 3   Rating                 1465 non-null   float64       
 4   Rating_count           1465 non-null   float64       
 5   Product_id             1465 non-null   object        
 6   Product_name           1465 non-null   object        
 7   Category               1465 non-null   object        
 8   About_product          1465 non-null   object        
 9   User_id                1465 non-null   object        
 10  User_name              1465 non-null   object        
 11  Review_id              1465 non-null   object        
 12  Review_title           1465 non-null   object        
 13  Rev

In [51]:
df2.to_csv('../data/cleaned/amazon.csv', index=False)
print(f'Successfully saved the new dataframe in a csv file format')

df2.head()

Successfully saved the new dataframe in a csv file format


Discounted_price_$  Actual_price_$  Discount_percentage_%  Rating   
0               399.0          1099.0                   64.0     4.2  \
1               199.0           349.0                   43.0     4.0   
2               199.0          1899.0                   90.0     3.9   
3               329.0           699.0                   53.0     4.2   
4               154.0           399.0                   61.0     4.2   

   Rating_count  Product_id   
0       24269.0  B07JW9H4J1  \
1       43994.0  B098NS6PVG   
2        7928.0  B096MSW6CT   
3       94363.0  B08HDJ86NZ   
4       16905.0  B08CF3B7N1   

                                        Product_name                 Category   
0  Wayona Nylon Braided USB to Lightning Fast Cha...    Computers&Accessories  \
1                                                 12  Accessories&Peripherals   
2                                                 11       Cables&Accessories   
3                                                  X                   Cables   
4                                                  8                USBCables   

                                    About_product   
0  High Compatibility : Compatible With iPhone 12  \
1                                              11   
2                                     X/XsMax/Xr    
3                                 iPhone 8/8 Plus   
4                                 iPhone 7/7 Plus   

                        User_id       User_name       Review_id   
0  AG3D6O4STAQKAY2UVGEUV46KN35Q           Manav  R3HXWT0LRP0NMF  \
1  AHMY5CWJMMK5BJRBBSNLYT3ONILA    Adarsh gupta  R2AJM3LFTLZHFO   
2  AHCTC6ULH4XB6YHDY6PCH2R772LQ         Sundeep    R6AQJGUP6P86   
3  AGYHHIERNXKA6P5T7CZLXKVPT7IQ  S.Sayeed Ahmed  R1KD19VHEDV0OR   
4  AG4OGOFWXJZTQ2HKYIOCOY3KXF2Q  jaspreet singh  R3C02RMYQMK6FC   

              Review_title                                     Review_content   
0                Satisfied     Looks durable Charging is fine tooNo complains  \
1  Charging is really fast                            Charging is really fast   
2          Value for money                                      good product.   
3           Product review               Till now satisfied with the quality.   
4             Good quality  This is a good product . The charging speed is...   

                                            Img_link   
0  https://m.media-amazon.com/images/W/WEBP_40237...  \
1  https://m.media-amazon.com/images/W/WEBP_40237...   
2  https://m.media-amazon.com/images/W/WEBP_40237...   
3  https://m.media-amazon.com/images/I/41V5FtEWPk...   
4  https://m.media-amazon.com/images/W/WEBP_40237...   

                                        Product_link  Revenue_$ Purchase_date   
0  https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...  1553216.0    2022-11-10  \
1  https://www.amazon.in/Ambrane-Unbreakable-Char...  1891742.0    2023-07-02   
2  https://www.amazon.in/Sounce-iPhone-Charging-C...   713520.0    2021-10-05   
3  https://www.amazon.in/Deuce-300-Resistant-Tang...  5001239.0    2023-12-22   
4  https://www.amazon.in/Portronics-Konnect-POR-1...  1031205.0    2023-06-18   

   Selling_price_$  Sales_volume Discount_bin  Price_change_%   
0            700.0       24269.0       60-70%      462.267553  \
1            150.0       43994.0       40-50%      -68.243858   
2           1700.0        7928.0       80-90%      444.126074   
3            370.0       94363.0       50-60%      -63.191153   
4            245.0       16905.0       60-70%      -42.918455   

   Quantity_change_%        PED  
0        2103.264801   4.549886  
1          81.276526  -1.190972  
2         -81.979361  -0.184586  
3        1090.249748 -17.253202  
4         -82.085139   1.912584

## Exploratory Data Analysis (EDA)

**Univariate analysis**

**Bivariate analysis**